connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java

```diff
@@ -0,0 +1,314 @@
+            adminClientConfig.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SSL");
+        }
+        return AdminClient.create(adminClientConfig);
+    }
+
+    /**
+     * Consume at least n records in a given duration or throw an exception.
+     *
+     * @param n the number of expected records in this topic.
+     * @param maxDuration the max duration to wait for these records.
+     * @param topics the topics to subscribe and consume records from.
+     * @return a {@link ConsumerRecords} collection containing at least n records.
+     */
+    public ConsumerRecords<byte[], byte[]> consume(int n, long maxDuration, String... topics) {
+        long num = 1;
+        long duration = maxDuration;
+        // poll in intervals of 250millis.
+        if (duration > 250) {
+            num = 1 + (maxDuration / 250);
+            duration = 250;
+        }
+
+        Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> records = new HashMap<>();
+        int consumedRecords = 0;
+        try (KafkaConsumer<byte[], byte[]> consumer = createConsumerAndSubscribeTo(topics)) {
+            for (int i = 0; i < num; i++) {
+                ConsumerRecords<byte[], byte[]> rec = consumer.poll(Duration.ofMillis(duration));
+                if (rec.isEmpty()) {
+                    continue;
+                }
+                for (TopicPartition partition: rec.partitions()) {
+                    final List<ConsumerRecord<byte[], byte[]>> r = rec.records(partition);
+                    records.computeIfAbsent(partition, t -> new ArrayList<>()).addAll(r);
+                    consumedRecords += r.size();
+                }
+                if (consumedRecords >= n) {
+                    return new ConsumerRecords<>(records);
+                }
+            }
+        }
+
+        throw new RuntimeException("Could not find enough records. found " + consumedRecords + ", expected " + n);
+    }
+
+    public KafkaConsumer<byte[], byte[]> createConsumer(Map<String, Object> consumerProps) {
+
+        Map<String, Object> props = new HashMap<>(consumerProps);
+
+        putIfAbsent(props, GROUP_ID_CONFIG, UUID.randomUUID().toString());
+        putIfAbsent(props, BOOTSTRAP_SERVERS_CONFIG, bootstrapServers());
+        putIfAbsent(props, ENABLE_AUTO_COMMIT_CONFIG, "false");
+        putIfAbsent(props, AUTO_OFFSET_RESET_CONFIG, "earliest");
+        putIfAbsent(props, KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        putIfAbsent(props, VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+
+        KafkaConsumer<byte[], byte[]> consumer;
+        try {
+            consumer = new KafkaConsumer<>(props);
+        } catch (Throwable t) {
+            throw new ConnectException("Failed to create consumer", t);
+        }
+        return consumer;
+    }
+
+    public KafkaConsumer<byte[], byte[]> createConsumer() {
```

Comment:
is this used, and if so, how? doesn't seem useful to be creating consumers that don't subscribe to anything

Reply:
this is used right below, line 312:

```
    public KafkaConsumer<byte[], byte[]> createConsumerAndSubscribeTo(String... topics) {
        KafkaConsumer<byte[], byte[]> consumer = createConsumer();
        consumer.subscribe(Arrays.asList(topics));
       ....
```

Reply:
right, my point was usage outside of this class. We have multiple public createConsumer* methods here, but the only one that's actually useful externally is createConsumerAndSubscribe. Being able to also override consumer props seems potentially useful, but that should just be two methods -- one that takes only topics and another that takes topics and consumer props overrides.